src/EFCore.PG/Query/NpgsqlSqlExpressionFactory.cs

```diff
@@ -138,6 +141,50 @@ RelationalTypeMapping FlipTimestampTypeMapping(RelationalTypeMapping mapping)
             [CanBeNull] RelationalTypeMapping typeMapping = null)
             => (PostgresNewArrayExpression)ApplyTypeMapping(new PostgresNewArrayExpression(initializers, type, typeMapping), typeMapping);
 
+        public override SqlBinaryExpression MakeBinary(
+            ExpressionType operatorType,
+            SqlExpression left,
+            SqlExpression right,
+            RelationalTypeMapping typeMapping)
+        {
+            Check.NotNull(left, nameof(left));
+            Check.NotNull(right, nameof(right));
+
+            if (operatorType == ExpressionType.Subtract)
+            {
+                if (left.Type == typeof(DateTime) && right.Type == typeof(DateTime) ||
+                    left.Type == typeof(DateTimeOffset) && right.Type == typeof(DateTimeOffset))
+                {
+                    return (SqlBinaryExpression)ApplyTypeMapping(
+                        new SqlBinaryExpression(operatorType, left, right, typeof(TimeSpan), null), typeMapping);
+                }
+
+                if (left.Type.FullName == "NodaTime.Instant" && right.Type.FullName == "NodaTime.Instant" ||
+                    left.Type.FullName == "NodaTime.ZonedDateTime" && right.Type.FullName == "NodaTime.ZonedDateTime")
+                {
+                    _nodaTimeDurationType ??= left.Type.Assembly.GetType("NodaTime.Duration");
```

Comment:
@smitpatel @bricelam @ajcvickers this is what happens to poor providers that try to do fancy things with plugins - not quite enough extension points for plugins...

Reply:
I think there are 2 orthogonal things here.
- Plugins can provide member/method translation but no other, specifically in this case operator "-".
   We used to support any custom expression in past because we had our own in relational at that point. Though there is huge perf penalty for that. Adding such a hook means before translating any expression in any form we have to consult through that rather than just restricting it to member/methods.
   We could bring back the same if we think we are fine with that perf penalty. Alternative could be adding a hook for plugins into pre/post translation where they can convert any custom expression they need to translate into a method wrapper and intercept that method instead.

- Second issue is SqlExpressionFactory. Plugins are not required to use SqlExpressionFactory. At the end of the day it is factory to create expressions with type mapping inferred. You could very well infer the types outside assign to sql expressions already and invoke this function without having any special logic here. TypeMappings are not reassigned when if already set. (Hence ISqlExpressionFactory exposes ApplyTypeMapping methods)

Reply:
> We used to support any custom expression in past because we had our own in relational at that point. Though there is huge perf penalty for that.

Yeah, mostly something to keep in mind - the current code with string-based lookups is ugly but it works fine. Maybe inside of a catch-all custom translation thing, we could simply have a hook where plugins can provide a SqlTranslatingEV, just like they can provide a method translator. Doesn't sound like it would be too bad for perf - just pretty niche.

> Plugins are not required to use SqlExpressionFactory.

One issue that I ran into, is that RelationalSqlTranslatingEV calls RelationalSqlExpressionFactory when translating binary operations, and by default the resulting SqlBinaryExpression's CLR type is the operands' (except for some bool cases). So for cases where the binary expression changes the CLR type (e.g. DateTime - DateTime => TimeSpan) you need to override MakeBinary or you get SqlBinaryExpressions with the wrong CLR type.

I get your point that we can bypass SqlExpressionFactory altogether.. But since these are operator translations (and not member/method), they can't be done in the plugin anyway, so they'd have to be done in NpgsqlSqlTranslatingEV anyway, which suffers from the same issue of not being to access NodaTime types. So I opted to just doing it the "nice way" via SqlExpressionFactory. It's also useful in case some other construct needs to be translated to one of these binary expressions - the logic would already be available for easy reuse.

Bottom line - I'm not complaining :) Just sharing what it currently looks like.

Reply:
> Maybe instead of a catch-all custom translation thing, we could simply have a hook where plugins can provide a SqlTranslatingEV

they are the same thing.

> So I opted to just doing it the "nice way" via SqlExpressionFactory.

That is the wrong way to do it. But if you want to do it that way, doesn't matter to core pipeline as long as it has facets it needs.

Reply:
@smitpatel can you provide more info on why you consider this the wrong way?

Operators are generally passed through as-is without needing custom translation logic, with the work being done between RelationalSqlTranslatingEV and SqlExpressionFactory. The only problem with these binary expressions is the type (and type mapping) changes, it seems better to me to solve these by handling the exceptional binary in SqlExpressionFactory (which already has specific logic for bool-returning binary expressions, for instance), rather than to add translation logic.

Also, this way the logic can get easily reused (e.g. a method translator somewhere can just create a regular binary expression with Instants and things will just work, rather than having to think about it there as well).

Reply:
The fragility of the code outweighs any benefit.
regarding reuse scenario, it is same as SqliteExpression.Strftime. You can refactor code to have a method to do any custom logic. That method does not have to be SqlExpressionFactory.MakeBinary.